### PR TITLE
chore: Move Agents into their own context, introduce Loadable abstraction

### PR DESCRIPTION
**Pull request by ClaireNeveu**
## Description
This is a PR @trentwatt and I put together. It does two things:

1. Split out Agents from the `Store` context. This is how we want to go about breaking up the Store context so that we can eventually switch to a more flexible way of managing global data.
2. Introduces a Loadable data type to manage loading states. Currently we're pretty bad at handling loading states and often just temporarily display incorrect data while loading. This will make it easier to implement loading states going forward and see where they're missing now.

Screen Shot 2022-10-28 at 14 50 23
Screen Shot 2022-10-28 at 15 09 42

## Test Plan
1. Visit the cluster overview page, confirm it matches latest-master (there are some features here I don't see in latest master if people have ideas on how to test them).
2. Confirm the navbar matches latest-master
4. Confirm the resource pool detail page matches latest-master
5. Replace each instance of `getAgents`, etc with `NotLoaded` and confirm the loading states display correctly.

## Checklist
- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: webui/react/src/components/Navigation.tsx
Comment (ClaireNeveu) [inline webui/react/src/components/Navigation.tsx:43]: This was previously happening inside a store action. This actually caused the effect to double up on one page. Makes more sense for it to live here.

File: webui/react/src/components/NavigationSideBar.tsx
Comment (ClaireNeveu) [inline webui/react/src/components/NavigationSideBar.tsx:119]: I couldn't find a spinner that actually worked in this spot so I left it a TODO. As a note: this label doesn't show up on latest-master which is annoying for testing.

File: webui/react/src/pages/Cluster/ClusterOverallBar.tsx
Comment (ClaireNeveu) [inline webui/react/src/pages/Cluster/ClusterOverallBar.tsx:14]: I honestly just couldn't figure out the correct loading state for this one (without writing a bunch of new CSS). Thought about having the whole page load but the loading state is only in one tab and I didn't see a way to have just the one tab loading (again, without writing a bunch of new code).

File: webui/react/src/pages/ResourcepoolDetail.tsx
Comment (ClaireNeveu) [inline webui/react/src/pages/ResourcepoolDetail.tsx:204]: Thanks to @keita-determined for linking me to some resources on how to use the new Suspense api in React 18. It makes some of these trickier loading states easier.

If you're wondering why I didn't use Suspense everywhere; it has a pretty big limitation in that a component can't manage its own loading state, you need to wrap the component usage in `<Suspense>` (and of course not all loading states should prevent rendering).

File: webui/react/src/stores/agents.tsx
Comment (ClaireNeveu) [inline webui/react/src/stores/agents.tsx:18]: Still used in the one place I couldn't sort out a good loading state.

File: webui/react/src/components/ResourcePoolCard.tsx
Comment (keita-determined) [inline webui/react/src/components/ResourcePoolCard.tsx:129]: Is this `Suspense` actually working?
I couldnt see `Spinner`

Comment (ClaireNeveu) [reply webui/react/src/components/ResourcePoolCard.tsx:129]: It is. When I was testing the loading states were so brief that I had to manually trigger them by adding `Loadable.waitFor(NotFound)` to see them.

Comment (keita-determined) [reply webui/react/src/components/ResourcePoolCard.tsx:129]: I tested it with `fast 3G` and `slow 3G`, but I didn't see spinner. 
Could you tell me how you tested with `Loadable.waitFor(NotFound)`? Prob im doing wrong

File: webui/react/src/pages/ResourcepoolDetail.tsx
Comment (keita-determined) [inline webui/react/src/pages/ResourcepoolDetail.tsx:204]: - in this case, I think `<Spinner />` would be more suitable
- Also, I couldnt see loading animation in this `Suspense`

File: webui/react/src/pages/ResourcepoolDetail.tsx
Comment (keita-determined) [inline webui/react/src/pages/ResourcepoolDetail.tsx:209]: I know its working with this name, but i think the name should be `ResourcepoolDetail` or adjust the name in https://github.com/determined-ai/determined/blob/bdf5b3ec5863e450ae0d9aa03d1fdee7588a7eac/webui/react/src/routes/index.tsx#L41
Then we can find the name much easier

File: webui/react/src/utils/loadable.ts
Comment (keita-determined) [inline webui/react/src/utils/loadable.ts:1]: im not very familiar with loadable, but i assume its similar to `Result` type in Rust and Swift. 
idk if we should write our own `Loadable` code. I think it would be used widely, so I think adding library is more reliable and more compatible with React in a long term.
I think we should talk about it before merging it (especially how we handle API calls)
I personally prefer to rely on popular and famous libraries as much as possible. Although this code could be perfect, someone else would change it in the future, and we might get regressions like `useSettings`.

File: webui/react/src/utils/loadable.ts
Comment (keita-determined) [inline webui/react/src/utils/loadable.ts:140]: `any` can be `unknown`, then `/* eslint-disable @typescript-eslint/no-explicit-any */` would be removed

File: webui/react/src/pages/ResourcepoolDetail.tsx
Comment (ClaireNeveu) [reply webui/react/src/pages/ResourcepoolDetail.tsx:209]: Sure, I can just rename `ResourcepoolDetail` to something else.

File: webui/react/src/utils/loadable.ts
Comment (ClaireNeveu) [reply webui/react/src/utils/loadable.ts:1]: It's better not to pull in libraries for things this small, it makes it difficult to audit the code and ensure we don't have security issues which is why I've never made Loadable into a library and just copied it around personal projects.

File: webui/react/src/pages/ResourcepoolDetail.tsx
Comment (ClaireNeveu) [reply webui/react/src/pages/ResourcepoolDetail.tsx]: Updated. I notice now that there's actually a skeleton for this page while it's loading. Can anybody help me figure out where that happens? That's where the suspense should live but I couldn't find the code that triggers that.

File: webui/react/src/pages/ResourcepoolDetail.tsx
Comment (ClaireNeveu) [reply webui/react/src/pages/ResourcepoolDetail.tsx]: Changed

File: webui/react/src/utils/loadable.ts
Comment (ClaireNeveu) [reply webui/react/src/utils/loadable.ts]: Thanks!

File: webui/react/src/utils/loadable.ts
Comment (keita-determined) [reply webui/react/src/utils/loadable.ts]: my concern is that if we add global state management tool, this code could be tech debt.
Also, React core team is talking about adding [`use` feature](https://github.com/reactjs/rfcs/pull/229) which is a similar purpose of usage. Even if its still under discussion, next.js v13 already added `use` func, so its likely to be added.

Comment (ClaireNeveu) [reply webui/react/src/utils/loadable.ts]: You don't have to worry about ADTs becoming tech debt, they're a fundamental programming technique.

File: webui/react/src/components/NavigationSideBar.tsx
Comment (trentwatt) [inline webui/react/src/components/NavigationSideBar.tsx:121]: cool!

File: webui/react/src/pages/Cluster/ClusterOverallBar.tsx
Comment (trentwatt) [reply webui/react/src/pages/Cluster/ClusterOverallBar.tsx:14]: I guess we could make `overview` a required prop and conditionally render this?

**Review by trentwatt — APPROVED** [review]
it LGTM but @hamidzr can you check out the netlify and/or chime in about anything in particular we should test?

the PR basically involves moving the cluster overview and resource pool details to its own context and changing slightly the fetch/loading patterns for their use.

File: webui/react/src/stores/agents.tsx
Comment (ClaireNeveu) [reply webui/react/src/stores/agents.tsx:96]: Yeah, the reason you don't want to have another store is because it's derived state. The agents list is the _source of truth_ for this data. Think of it the same way you think of a database, you want to avoid having the same information in multiple places because they could get out of sync.

File: webui/react/src/stores/agents.tsx
Comment (ClaireNeveu) [reply webui/react/src/stores/agents.tsx:62]: Your list is accurate but in my experience you want to give components control over the freshness of data. In this particular case you don't need to but I want the stores to have parity of functionality so it's easier to work with them.

File: webui/react/src/stores/agents.tsx
Comment (trentwatt) [reply webui/react/src/stores/agents.tsx:96]: i was thinking to have in the same store, just as a different part of it. 

analogy to database would be that you have a materialized view (derived state) that is kept in sync with the underlying data and then accessed frequently.

Comment (ClaireNeveu) [reply webui/react/src/stores/agents.tsx:96]: I would say that `useMemo` is the analogue for a materialized view.

File: webui/react/src/stores/agents.tsx
Comment (trentwatt) [reply webui/react/src/stores/agents.tsx:62]: (Feel free to ignore this one if it is getting too far afield for this PR)

what about something like `useAgents` and `useFreshAgents`?  (okay the name is pretty bad).

at that point the functionality would be basically the same as existing `useFetchAgents`/`useEnsureAgentsFetched` (i think)

but then if the calls happen within the provider itself it could be possible to make it so that the fetch only happens once even if `useFreshAgents` is happening in multiple places.

File: webui/react/src/stores/agents.tsx
Comment (trentwatt) [reply webui/react/src/stores/agents.tsx:96]: in this case for `useMemo`, though, the view is materialized separately for each client consuming it.

File: webui/react/src/stores/agents.tsx
Comment (ClaireNeveu) [reply webui/react/src/stores/agents.tsx:62]: So the issue there is you're talking about the "getting" side which shouldn't know about how the data arrives. Neither `useAgents` nor `useFreshAgents` will do anything unless an appropriate action is triggered somewhere else.

File: webui/react/src/stores/agents.tsx
Comment (ClaireNeveu) [reply webui/react/src/stores/agents.tsx:96]: That's specific to this implementation but it needn't be. We could put the `useMemo` in the context provider.

Comment (ClaireNeveu) [reply webui/react/src/stores/agents.tsx:96]: Sorry, misunderstood you. I think that's a good idea if the computation is expensive. I think here it's probably not worth the extra complexity because the computation is so cheap (unless people have a lot more agents than I'm imagining)

Comment (trentwatt) [reply webui/react/src/stores/agents.tsx:96]: fair enough. agree on the cheapness of the computation.

**Review by hamidzr — COMMENTED** [review]
> it LGTM but @hamidzr can you check out the netlify and/or chime in about anything in particular we should test?

I quickly checked out the changes LG. One scenario that might help to check is when we go from no agents to some connected agents and vice-versa. 

- one loading-related issue with rp card story



not introduced here:
- I noticed the sidebar has a loading state.  ideally, nav would up before pieces that need to load within it (workspace entries)

File: webui/react/src/pages/ResourcepoolDetail.tsx
Comment (ClaireNeveu) [reply webui/react/src/pages/ResourcepoolDetail.tsx]: Okay, I finally had a chance to look into this. So the reason the suspense here is never hit is because of timing around different API calls but it's not _inherently_ inaccessible. Once we convert more data management to use this new pattern we should see the suspense being hit.

**Pull request merged**
